hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapOperation.java

```diff
@@ -125,16 +126,34 @@ public final void beforeRun() throws Exception {
                 && recordStore != null
                 && recordStore.getMapDataStore() != MapDataStores.EMPTY_MAP_DATA_STORE);
 
-        // check mapStoreOffloadEnabled is true for current operation
-        mapStoreOffloadEnabled = recordStore != null && hasUserConfiguredOffload
+        // check if mapStoreOffloadEnabled is true for this operation
+        mapStoreOffloadEnabled = recordStore != null
+                && hasUserConfiguredOffload
                 && getStartingStep() != null
                 && !mapConfig.getTieredStoreConfig().isEnabled();
 
+        // check if tieredStore and partitionCompactor
+        // are both enabled for this operation
+        tieredStoreAndPartitionCompactorEnabled = recordStore != null
+                && recordStore.getStorage().isPartitionCompactorEnabled()
+                && getStartingStep() != null
+                && mapConfig.getTieredStoreConfig().isEnabled();
+
+        assertOnlyOneOfMapStoreOrTieredStoreEnabled();
         assertNativeMapOnPartitionThread();
 
         innerBeforeRun();
     }
 
+    private void assertOnlyOneOfMapStoreOrTieredStoreEnabled() {
+        if (!ASSERTION_ENABLED) {
+            return;
+        }
```

Comment:
I don't get this `if`. If the assertions are disabled the entire `assert` should be eliminated, and no need for the `if`. Do I miss something?

Reply:
Inlined it what we have currently int that class but probably no need for it, it looks like a micro optimization attempt.